src/hardware/shared_adc.rs

```diff
@@ -0,0 +1,117 @@
+/// Shared Internal ADC Support
+///
+/// # Description
+/// This module provides an abstraction to share ownership of a single ADC peripheral with multiple
+/// ADC channels attached to it.
+///
+/// The design of this module mimics that of [`shared-bus`].
+///
+/// First, the shared ADC is created with the use of a macro, which places the ADC peripheral into
+/// a mutable, static (singleton) location. Then, individual channels are created by passing in the
+/// associated ADC input pin to the [SharedAdc::create_channel()] function to generate an
+/// [AdcChannel]. The [AdcChannel]'s ownership can then be moved to any required drivers.
+///
+/// ## Synchronization
+/// Currently, the sharing of the ADC peripheral across multiple priority levels is managed by the
+/// use of a critical section.
+use embedded_hal::adc::{Channel, OneShot};
+use stm32h7xx_hal as hal;
+
+#[derive(Debug, Copy, Clone)]
+pub enum AdcError {
+    /// Indicates that the ADC channel has already been allocated.
+    Allocated,
+}
+
+/// A single channel on an ADC peripheral.
+pub struct AdcChannel<'a, Adc, PIN> {
+    pin: PIN,
+    slope: f32,
```

Comment:
also add offset such that: `read() -> raw_voltage * slope + offset`

Reply:
The `read()` function is really only intended to provide a normalized reading (e.g. 0-1 ADC scale). The slope here is the `adcX.slope()` result, which indicates the upper-limit of measurement.

Reply:
Ack.